### PR TITLE
[dsub] Implement querying by end date

### LINKED ISSUE #73: Support querying by start and end times
- [x] Query by `start` (actually done by create not submission https://github.com/bvprivate/job-monitor/issues/54)
- [ ] Query by `end` (client side until supported by dsub)
- [ ] Query by `last-update` (client side)

### PR DESCRIPTION
**Pull request by bfcrampton**
### Resolves #73:
- Implement querying by end datetime
- Minor cleanup in tests to shorten a couple function names and make it more readable

### CODE REVIEW

File: servers/dsub/jobs/controllers/jobs_controller.py [hunk 1]
```diff
@@ -129,14 +129,21 @@ def query_jobs(body):
...
+            raise BadRequest(
+                "Invalid query: start date must precede end date.")
 
     job_generator = _generate_dstat_jobs(provider, query, create_time_max,
                                          offset_id)
     jobs = []
     try:
         for job in job_generator:
-            jobs.append(job)
+            if not query.end or job.get(
```
Comment (calbach) [inline servers/dsub/jobs/controllers/jobs_controller.py]: nit: I would probably invert and `continue` instead, which should set up better for additional filter criteria.

Comment (calbach) [inline servers/dsub/jobs/controllers/jobs_controller.py]: I think you want this inside the job generator function instead. Consider the following test case:

request: pageSize of 2, end=10
jobs: {end: 4}, {end: 5}, {end: 11}

want: page of 2 elements
got: page of 2 elements, page of 0 elements

**Comment by calbach** [comment]
Per offline discussion, should also set `create_time_max` on the initial request if `query.end` is given.

File: servers/dsub/jobs/controllers/jobs_controller.py (hunk 1 above)
Comment (bfcrampton) [reply servers/dsub/jobs/controllers/jobs_controller.py]: Done (in generator function)

Comment (bfcrampton) [reply servers/dsub/jobs/controllers/jobs_controller.py]: Good point didn't think about it returning page tokens when it shouldn't. Fixed.

File: servers/dsub/jobs/controllers/jobs_controller.py [hunk 2]
```diff
@@ -164,6 +169,12 @@ def _auth_token():
 
 def _generate_dstat_jobs(provider, query, create_time_max=None,
                          offset_id=None):
+    # If create_time_max is not set, but we have to client-side filter by
+    # end-time, set create_time_max = query.end because all the jobs with
+    # create_time >= query.end cannot possibly match the query.
+    if not create_time_max and query.start:
```
Comment (calbach) [inline servers/dsub/jobs/controllers/jobs_controller.py]: hmmm.. I thought this was supposed to be query.end?

Do the tests fail?

File: servers/dsub/jobs/controllers/jobs_controller.py [hunk 3]
```diff
@@ -179,6 +190,9 @@ def _generate_dstat_jobs(provider, query, create_time_max=None,
     last_create_time = None
     job_buffer = []
     for job in jobs:
+        if query.end and job['end-time'] > query.end:
```
Comment (calbach) [inline servers/dsub/jobs/controllers/jobs_controller.py]: Wouldn't this fail for running jobs, i.e. without an end time?

File: servers/dsub/jobs/controllers/jobs_controller.py (hunk 2 above)
Comment (bfcrampton) [reply servers/dsub/jobs/controllers/jobs_controller.py]: Whoops you're definitely should be `query.end`, should've run the tests again to make sure they still pass.

File: servers/dsub/jobs/controllers/jobs_controller.py (hunk 3 above)
Comment (bfcrampton) [reply servers/dsub/jobs/controllers/jobs_controller.py]: Yup, fixed.

**Pull request merged**